src/net/dns.rs

```diff
@@ -42,33 +82,105 @@ pub(crate) async fn prune_dns_cache(context: &Context) -> Result<()> {
     Ok(())
 }
 
-/// Looks up the hostname and updates DNS cache
-/// on success.
+/// Map from hostname to IP addresses.
+static LOOKUP_HOST_CACHE: Lazy<std::sync::Mutex<HashMap<String, Vec<IpAddr>>>> =
+    Lazy::new(Default::default);
+
+async fn lookup_host_with_memory_cache(
+    context: &Context,
+    hostname: &str,
+    port: u16,
+) -> Result<Vec<IpAddr>> {
+    let stale_result = LOOKUP_HOST_CACHE.lock().unwrap().get(hostname).cloned();
```

Comment:
This line looks a bit dangerous. I know, the MutexGuard is dropped in this line, but this is not as obvious as it should be, esp. since holding the MutexGuard across an await could lead to deadlocks.

I would prefer to either extract this line into its own function, e.g.:

```rust
fn lookup_host_cache(hostname: &str) -> Option<Vec<IpAddr>> {
    LOOKUP_HOST_CACHE.lock().unwrap().get(hostname).cloned()
}
```

(in this case, would be nice to also create a `fn insert_into_host_cache(..)` for symmetry)

or to drop it explicitly:

```rust
let mutex_guard = LOOKUP_HOST_CACHE.lock().unwrap();
let stale_result = mutex_guard.get(hostname).cloned();
drop(mutex_guard);
```

Reply:
The fact that 

```rust
    let mutex_guard = LOOKUP_HOST_CACHE.lock().unwrap();
    let stale_result = mutex_guard.get(hostname).cloned();
    drop(mutex_guard);
```

doesn't work here really surprises me, and there seems to be an incorrectness in my mental model of rust async . We are dropping mutex_guard, so why is it still alive in line 135?

My assumption is that it's because the compiler still tries to put the `mutex_guard` place into the async state machine, even though it's dropped (i.e. logically uninitialized memory) when the `.await` is there.

According to https://www.reddit.com/r/rust/comments/vadj90/dropping_mutex_guards/, this code would have been correct in the way that the mutex would have been correctly unlocked.

On the upside, it's good to have the reassurance that the compiler would catch us writing

```rust
    let mutex_guard = LOOKUP_HOST_CACHE.lock().unwrap();
    let stale_result = mutex_guard.get(hostname).cloned();
```
i.e. not dropping the mutex_guard and then later holding it over an `.await` point.

Anyway, your solution is nice, [and doesn't exhibit the usual footgun of MutexGuards not being dropped correctly](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6c46447f4ed8630d6f4ead5996b365ea).